src/AspNetCoreModuleV2/AspNetCore/src/applicationinfo.cpp

```diff
@@ -98,73 +143,45 @@ APPLICATION_INFO::UpdateAppOfflineFileHandle()
     UTILITY::ConvertPathToFullPath(L".\\app_offline.htm",
         m_pConfiguration->QueryApplicationPhysicalPath()->QueryStr(),
         &strFilePath);
-    APP_OFFLINE_HTM *pOldAppOfflineHtm = NULL;
-    APP_OFFLINE_HTM *pNewAppOfflineHtm = NULL;
-
-    ReferenceApplicationInfo();
 
     if (INVALID_FILE_ATTRIBUTES == GetFileAttributes(strFilePath.QueryStr()))
     {
-        // Check if app offline was originally present.
-        // if it was, log that app_offline has been dropped.
-        if (m_fAppOfflineFound)
-        {
-            UTILITY::LogEvent(g_hEventLog,
-                EVENTLOG_INFORMATION_TYPE,
-                ASPNETCORE_EVENT_RECYCLE_APPOFFLINE_REMOVED,
-                ASPNETCORE_EVENT_RECYCLE_APPOFFLINE_REMOVED_MSG);
-        }
-
         m_fAppOfflineFound = FALSE;
     }
     else
     {
-        pNewAppOfflineHtm = new APP_OFFLINE_HTM(strFilePath.QueryStr());
-
-        if (pNewAppOfflineHtm != NULL)
+        // Only load the appoffline file once to avoid expensive IO
+        // If appoffline file changes during offline, we will not refresh it
+        if (!m_fAppOfflineLoaded)
         {
-            if (pNewAppOfflineHtm->Load())
-            {
-                //
-                // loaded the new app_offline.htm
-                //
-                pOldAppOfflineHtm = (APP_OFFLINE_HTM *)InterlockedExchangePointer((VOID**)&m_pAppOfflineHtm, pNewAppOfflineHtm);
-
-                if (pOldAppOfflineHtm != NULL)
-                {
-                    pOldAppOfflineHtm->DereferenceAppOfflineHtm();
-                    pOldAppOfflineHtm = NULL;
-                }
-            }
-            else
-            {
-                // ignored the new app_offline file because the file does not exist.
-                pNewAppOfflineHtm->DereferenceAppOfflineHtm();
-                pNewAppOfflineHtm = NULL;
-            }
-        }
-
-        m_fAppOfflineFound = TRUE;
-
-        // recycle the application
-        if (m_pApplication != NULL)
-        {
-            STACK_STRU(strEventMsg, 256);
-            if (SUCCEEDED(strEventMsg.SafeSnwprintf(
-                ASPNETCORE_EVENT_RECYCLE_APPOFFLINE_MSG,
-                m_pConfiguration->QueryApplicationPath()->QueryStr())))
-            {
-                UTILITY::LogEvent(g_hEventLog,
-                    EVENTLOG_INFORMATION_TYPE,
-                    ASPNETCORE_EVENT_RECYCLE_APPOFFLINE,
-                    strEventMsg.QueryStr());
-            }
-
-            RecycleApplication();
+            LoadAppOffline(strFilePath.QueryStr());
```

Comment:
You never check the return value of LoadAppOffline(). Doesn't this determine m_fAppOfflineFound ?

Reply:
we may fail to load the app_offline content due to file ACL or the file is exclusively opened by some other proces (e.g., antivirus) but the file does exist. this is why we set m_fAppOfflineFound to true. We will return 503 with empty content

Reply:
Why does it have return value at all then?